File: python/megane/protocol.py
"""Binary protocol for Python -> JavaScript data transfer.

Protocol format:
  Header (8 bytes):
    magic:    4 bytes "MEGN"
    msg_type: u8 (0=snapshot, 1=frame, 2=metadata)
    flags:    u8 (bit 0: HAS_BOND_ORDERS, bit 1: HAS_BOX, bit 3: HAS_BOX_ORIGIN,
                  bit 4: HAS_SYMMETRY_OPS)
    reserved: 2 bytes

  Snapshot payload:
    n_atoms:      u32
    n_bonds:      u32
    positions:    Float32[n_atoms * 3]
    elements:     Uint8[n_atoms]     + pad to 4 bytes
    bonds:        Uint32[n_bonds * 2]
    bond_orders:  Uint8[n_bonds]     + pad to 4 bytes  (if HAS_BOND_ORDERS)
    box:          Float32[9]         (3x3 row-major)   (if HAS_BOX)
    box_origin:   Float32[3]         (xlo,ylo,zlo)     (if HAS_BOX_ORIGIN)
    symmetry_ops: u32 byte length + newline-joined utf-8 `x,y,z` strings
                  + pad to 4 bytes                     (if HAS_SYMMETRY_OPS)

  Frame payload:
    frame_id:   u32
    n_atoms:    u32
    positions:  Float32[n_atoms * 3]
    elements:   Uint8[n_atoms]  + pad to 4 bytes  (if HAS_FRAME_ELEMENTS)
    box:        Float32[9]       (3x3 row-major)   (if HAS_BOX)

  The optional frame elements/box carry per-frame topology/cell for
  heterogeneous trajectories; a frame that omits them (flags 0) is byte-identical
  to the original positions-only layout, so uniform playback is unchanged.
"""

from __future__ import annotations

import struct
from typing import Protocol, runtime_checkable

import numpy as np

__all__ = [
    "encode_snapshot",
    "encode_frame",
    "encode_metadata",
    "StructureLike",
    "MSG_SNAPSHOT",
    "MSG_FRAME",
    "MSG_METADATA",
]

MAGIC = b"MEGN"
MSG_SNAPSHOT = 0
MSG_FRAME = 1
MSG_METADATA = 2

HAS_BOND_ORDERS = 0x01
HAS_BOX = 0x02
HAS_FRAME_ELEMENTS = 0x04
HAS_BOX_ORIGIN = 0x08
HAS_SYMMETRY_OPS = 0x10


@runtime_checkable
class StructureLike(Protocol):
    """Structural contract for objects that can be encoded as snapshots."""

    n_atoms: int
    positions: np.ndarray  # (N, 3) float32
    elements: np.ndarray  # (N,) uint8
    bonds: np.ndarray  # (M, 2) uint32
    bond_orders: np.ndarray  # (M,) uint8
    box: np.ndarray  # (3, 3) float32
    box_origin: np.ndarray  # (3,) float32 — box lower corner; zero if unset


def encode_snapshot(structure: StructureLike) -> bytes:
    """Encode a molecular structure as a binary snapshot message."""
    n_atoms = structure.n_atoms
    n_bonds = len(structure.bonds)

    flags = 0

    # Positions: float32 * n_atoms * 3
    pos_bytes = structure.positions.astype(np.float32).tobytes()

    # Elements: uint8 * n_atoms, padded to 4-byte alignment
    elem_bytes = structure.elements.astype(np.uint8).tobytes()
    padding_len = (4 - (len(elem_bytes) % 4)) % 4
    elem_bytes += b"\x00" * padding_len

    # Bonds: uint32 * n_bonds * 2
    bond_bytes = structure.bonds.astype(np.uint32).tobytes()

    # Bond orders (optional)
    bond_order_bytes = b""
    if len(structure.bond_orders) > 0:
        flags |= HAS_BOND_ORDERS
        bo_raw = structure.bond_orders.astype(np.uint8).tobytes()
        bo_padding = (4 - (len(bo_raw) % 4)) % 4
        bond_order_bytes = bo_raw + b"\x00" * bo_padding

    # Box (optional)
    box_bytes = b""
    if np.any(structure.box != 0):
        flags |= HAS_BOX
        box_bytes = structure.box.astype(np.float32).flatten().tobytes()

    # Box origin (optional; only meaningful alongside a box). Emitted when the
    # lower corner is offset from the world origin — e.g. a LAMMPS box far from
    # (0,0,0) — so the frontend draws the cell around its (absolute) atoms.
    origin_bytes = b""
    box_origin = getattr(structure, "box_origin", None)
    if (flags & HAS_BOX) and box_origin is not None and np.any(np.asarray(box_origin) != 0):
        flags |= HAS_BOX_ORIGIN
        origin_bytes = np.asarray(box_origin, dtype=np.float32).flatten().tobytes()

    # Symmetry operations (optional): u32 byte length + newline-joined utf-8
    # `x,y,z` strings, padded to 4-byte alignment. The frontend symmetry node
    # applies them; the snapshot itself stays the asymmetric unit as parsed.
    symops_bytes = b""
    symmetry_ops = getattr(structure, "symmetry_ops", None)
    if symmetry_ops:
        flags |= HAS_SYMMETRY_OPS
        raw = "\n".join(symmetry_ops).encode("utf-8")
        sym_padding = (4 - ((4 + len(raw)) % 4)) % 4
        symops_bytes = struct.pack("<I", len(raw)) + raw + b"\x00" * sym_padding

    # Header: magic(4) + msg_type(1) + flags(1) + reserved(2) = 8 bytes
    header = MAGIC + struct.pack("<BBH", MSG_SNAPSHOT, flags, 0)

    # Snapshot header: n_atoms(4) + n_bonds(4) = 8 bytes
    snapshot_header = struct.pack("<II", n_atoms, n_bonds)

    return (
        header
        + snapshot_header
        + pos_bytes
        + elem_bytes
        + bond_bytes
        + bond_order_bytes
        + box_bytes
        + origin_bytes
        + symops_bytes
    )


def encode_frame(
    frame_id: int,
    positions: np.ndarray,
    elements: np.ndarray | None = None,
    box: np.ndarray | None = None,
) -> bytes:
    """Encode a trajectory frame as a binary message.

    Args:
        frame_id: 0-based frame index.
        positions: (N, 3) float32 atom positions.
        elements: optional (N,) uint8 per-frame atomic numbers (heterogeneous
            topology). Omit for uniform trajectories (host reuses the snapshot).
        box: optional (3, 3) float32 per-frame unit cell. Omit for a constant cell.

    A call with neither ``elements`` nor ``box`` produces the original
    positions-only layout, so uniform playback is byte-identical.
    """
    n_atoms = len(positions)

    flags = 0
    pos_bytes = positions.astype(np.float32).tobytes()

    elem_bytes = b""
    if elements is not None:
        flags |= HAS_FRAME_ELEMENTS
        raw = np.asarray(elements, dtype=np.uint8).tobytes()
        pad = (4 - (len(raw) % 4)) % 4
        elem_bytes = raw + b"\x00" * pad

    box_bytes = b""
    if box is not None:
        flags |= HAS_BOX
        box_bytes = np.asarray(box, dtype=np.float32).flatten().tobytes()

    header = MAGIC + struct.pack("<BBH", MSG_FRAME, flags, 0)
    frame_header = struct.pack("<II", frame_id, n_atoms)

    return header + frame_header + pos_bytes + elem_bytes + box_bytes


def encode_metadata(
    n_frames: int,
    timestep_ps: float,
    n_atoms: int,
    pdb_name: str = "",
    xtc_name: str = "",
) -> bytes:
    """Encode trajectory metadata message.

    Includes optional file name strings appended as length-prefixed UTF-8.
    """
    header = MAGIC + struct.pack("<BBH", MSG_METADATA, 0, 0)
    payload = struct.pack("<IfI", n_frames, timestep_ps, n_atoms)
    # Append file names as length-prefixed UTF-8
    pdb_bytes = pdb_name.encode("utf-8")
    xtc_bytes = xtc_name.encode("utf-8")
    payload += struct.pack("<H", len(pdb_bytes)) + pdb_bytes
    payload += struct.pack("<H", len(xtc_bytes)) + xtc_bytes
    return header + payload

File: src/protocol/protocol.ts
/**
 * Binary protocol decoder for megane.
 *
 * Format matches python/megane/protocol.py.
 * Header: "MEGN" (4 bytes) + msg_type (u8) + flags (u8) + reserved (u16)
 */

import type { Snapshot, Frame, TrajectoryMeta } from "../types";

const MAGIC = 0x4e47454d; // "MEGN" in little-endian

export const MSG_SNAPSHOT = 0;
export const MSG_FRAME = 1;
export const MSG_METADATA = 2;

const HAS_BOND_ORDERS = 0x01;
const HAS_BOX = 0x02;
const HAS_FRAME_ELEMENTS = 0x04;
const HAS_BOX_ORIGIN = 0x08;
const HAS_SYMMETRY_OPS = 0x10;

export function decodeHeader(buffer: ArrayBuffer): {
  msgType: number;
  flags: number;
} {
  const view = new DataView(buffer);
  const magic = view.getUint32(0, true);
  if (magic !== MAGIC) {
    throw new Error(`Invalid magic: 0x${magic.toString(16)}, expected 0x${MAGIC.toString(16)}`);
  }
  return {
    msgType: view.getUint8(4),
    flags: view.getUint8(5),
  };
}

export function decodeSnapshot(buffer: ArrayBuffer): Snapshot {
  const view = new DataView(buffer);
  const flags = view.getUint8(5);
  let offset = 8; // skip header

  const nAtoms = view.getUint32(offset, true);
  offset += 4;
  const nBonds = view.getUint32(offset, true);
  offset += 4;

  // Positions: Float32Array[nAtoms * 3]
  const positions = new Float32Array(buffer, offset, nAtoms * 3);
  offset += nAtoms * 3 * 4;

  // Elements: Uint8Array[nAtoms], padded to 4-byte alignment
  const elements = new Uint8Array(buffer, offset, nAtoms);
  offset += nAtoms;
  offset += (4 - (offset % 4)) % 4; // alignment padding

  // Bonds: Uint32Array[nBonds * 2]
  const bonds = new Uint32Array(buffer, offset, nBonds * 2);
  offset += nBonds * 2 * 4;

  // Bond orders (optional, if HAS_BOND_ORDERS flag)
  let bondOrders: Uint8Array | null = null;
  if (flags & HAS_BOND_ORDERS) {
    bondOrders = new Uint8Array(buffer, offset, nBonds);
    offset += nBonds;
    offset += (4 - (offset % 4)) % 4; // alignment padding
  }

  // Box (optional, if HAS_BOX flag)
  let box: Float32Array | null = null;
  if (flags & HAS_BOX) {
    box = new Float32Array(buffer, offset, 9);
    offset += 9 * 4;
  }

  // Box origin (optional, if HAS_BOX_ORIGIN flag) — follows the box.
  let boxOrigin: Float32Array | null = null;
  if (flags & HAS_BOX_ORIGIN) {
    boxOrigin = new Float32Array(buffer, offset, 3);
    offset += 3 * 4;
  }

  // Symmetry operations (optional, if HAS_SYMMETRY_OPS flag): u32 byte length
  // + newline-joined utf-8 `x,y,z` strings. The snapshot stays the asymmetric
  // unit; the pipeline's symmetry node applies these downstream.
  let symmetryOps: string[] | undefined;
  if (flags & HAS_SYMMETRY_OPS) {
    const byteLen = view.getUint32(offset, true);
    offset += 4;
    const raw = new Uint8Array(buffer, offset, byteLen);
    symmetryOps = new TextDecoder().decode(raw).split("\n");
    offset += byteLen;
    offset += (4 - (offset % 4)) % 4; // alignment padding
  }

  return {
    nAtoms,
    nBonds,
    nFileBonds: nBonds,
    positions,
    elements,
    bonds,
    bondOrders,
    box,
    boxOrigin,
    atomChainIds: null,
    atomBFactors: null,
    ...(symmetryOps && { symmetryOps }),
  };
}

export function decodeFrame(buffer: ArrayBuffer): Frame {
  const view = new DataView(buffer);
  const flags = view.getUint8(5);
  let offset = 8; // skip header

  const frameId = view.getUint32(offset, true);
  offset += 4;
  const nAtoms = view.getUint32(offset, true);
  offset += 4;

  const positions = new Float32Array(buffer, offset, nAtoms * 3);
  offset += nAtoms * 3 * 4;

  // Optional per-frame elements (heterogeneous topology), padded to 4 bytes.
  let elements: Uint8Array | undefined;
  if (flags & HAS_FRAME_ELEMENTS) {
    elements = new Uint8Array(buffer, offset, nAtoms);
    offset += Math.ceil(nAtoms / 4) * 4;
  }

  // Optional per-frame unit cell (heterogeneous cell).
  let box: Float32Array | undefined;
  if (flags & HAS_BOX) {
    box = new Float32Array(buffer, offset, 9);
    offset += 9 * 4;
  }

  return { frameId, nAtoms, positions, ...(elements ? { elements } : {}), ...(box ? { box } : {}) };
}

export function decodeMetadata(buffer: ArrayBuffer): TrajectoryMeta {
  const view = new DataView(buffer);
  const offset = 8; // skip header

  const nFrames = view.getUint32(offset, true);
  const timestepPs = view.getFloat32(offset + 4, true);
  const nAtoms = view.getUint32(offset + 8, true);

  let pdbName: string | undefined;
  let xtcName: string | undefined;

  // Parse file names if buffer has extended data (backward compatible)
  if (buffer.byteLength > 20) {
    const decoder = new TextDecoder();
    let pos = 20;
    const pdbLen = view.getUint16(pos, true);
    pos += 2;
    if (pdbLen > 0) {
      pdbName = decoder.decode(new Uint8Array(buffer, pos, pdbLen));
    }
    pos += pdbLen;
    const xtcLen = view.getUint16(pos, true);
    pos += 2;
    if (xtcLen > 0) {
      xtcName = decoder.decode(new Uint8Array(buffer, pos, xtcLen));
    }
  }

  return { nFrames, timestepPs, nAtoms, pdbName, xtcName };
}
